parsers/OPENNEM.py

```diff
@@ -0,0 +1,216 @@
+    # Require all present keys to be non-null
+    if filtered_keys and pd_series[filtered_keys].notnull().all():
+        return pd_series[filtered_keys].apply(transform).sum()
+    else:
+        return None
+
+
+def fetch_production(zone_key=None, session=None, target_datetime=None, logger=logging.getLogger(__name__)):
+    df, capacities = fetch_main_df('power', zone_key=zone_key, session=session, target_datetime=target_datetime, logger=logger)
+    # Drop interconnectors
+    df = df.drop([x for x in df.columns if '->' in x], axis=1)
+
+    # Make sure charging is counted positively
+    # and discharging negetively
+    if 'BATTERY_DISCHARGING' in df.columns:
+        df['BATTERY_DISCHARGING'] = df['BATTERY_DISCHARGING'] * -1
+
+    objs = [{
+        'datetime': arrow.get(dt.to_pydatetime()).datetime,
+        'production': {  # Unit is MW
+            'coal': sum_vector(row, OPENNEM_PRODUCTION_CATEGORIES['coal']),
+            'gas': sum_vector(row, OPENNEM_PRODUCTION_CATEGORIES['gas']),
+            'oil': sum_vector(row, OPENNEM_PRODUCTION_CATEGORIES['oil']),
+            'hydro': sum_vector(row, OPENNEM_PRODUCTION_CATEGORIES['hydro']),
+            'wind': sum_vector(row, OPENNEM_PRODUCTION_CATEGORIES['wind']),
+            'biomass': sum_vector(row, OPENNEM_PRODUCTION_CATEGORIES['biomass']),
+            # We here assume all rooftop solar is fed to the grid
+            # This assumption should be checked and we should here only report
+            # grid-level generation
+            'solar': sum_vector(row, OPENNEM_PRODUCTION_CATEGORIES['solar']),
+        },
+        'storage': {
+            # opennem reports charging as negative, we here should report as positive
+            # Note: we made the sign switch before, so we can just sum them up
+            'battery': sum_vector(row, OPENNEM_STORAGE_CATEGORIES['battery']),
+            # opennem reports pumping as positive, we here should report as positive
+            'hydro': sum_vector(row, OPENNEM_STORAGE_CATEGORIES['hydro']),
+        },
+        'capacity': {
+            'coal': sum_vector(capacities, OPENNEM_PRODUCTION_CATEGORIES['coal']),
+            'gas': sum_vector(capacities, OPENNEM_PRODUCTION_CATEGORIES['gas']),
+            'oil': sum_vector(capacities, OPENNEM_PRODUCTION_CATEGORIES['oil']),
+            'hydro': sum_vector(capacities, OPENNEM_PRODUCTION_CATEGORIES['hydro']),
+            'wind': sum_vector(capacities, OPENNEM_PRODUCTION_CATEGORIES['wind']),
+            'biomass': sum_vector(capacities, OPENNEM_PRODUCTION_CATEGORIES['biomass']),
+            'solar': sum_vector(capacities, OPENNEM_PRODUCTION_CATEGORIES['solar']),
+        },
+        'source': SOURCE,
+        'zoneKey': zone_key,
+    } for dt, row in df.iterrows()]
+
+    # Validation
+    for obj in objs:
+        for k, v in obj['production'].items():
+            if v is None:
+                continue
+            if v < 0 and v > -50:
+                # Set small negative values to 0
```

Comment:
Wondering why?

Reply:
Sometimes certain power plants can draw a bit of electricity from the grid just to stay online (even though they produce "nothing"). We currently don't support having negative production so we simply set to 0.